File: SCL/Sort/Quicksort/quickSort.py
from typing import List

def quickSort(q:List,l:int,r:int):
    if l >= r:
        return
    i,j,x = l - 1,r + 1,q[l]
    while i < j:
        i += 1
        while q[i] < x:
            i += 1
        j -= 1
        while q[j] > x:
            j -= 1
        if i < j:
            q[i],q[j] = q[j],q[i]
    quickSort(q,l,j)
    quickSort(q,j + 1,r)

if __name__ == "__main__":
    input = [2,1,2,4,5]
    quickSort(input,0,4)
    for i in input:
        print(i)

File: Sort/Quicksort/quickSort.go
package main

import "fmt"

func quickSort(input []int, l, r int) {
	if l >= r {
		return
	}
	i, j, x := l-1, r+1, input[l]
	for i < j {
		i++
		for input[i] < x {
			i++
		}
		j--
		for input[j] > x {
			j--
		}
		if i < j {
			input[i], input[j] = input[j], input[i]
		}
	}
	quickSort(input, l, j)
	quickSort(input, j+1, r)
}

func main() {
	input := []int{
		2, 2, 1, 4, 5}
	quickSort(input, 0, len(input)-1)
	for i := range input {
		fmt.Println(input[i])
	}
}
